File: symmetric/symmetric-core/src/main/java/org/jumpmind/symmetric/service/impl/PurgeService.java
/*
 * Licensed to JumpMind Inc under one or more contributor 
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding 
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU Lesser General Public License (the
 * "License"); you may not use this file except in compliance
 * with the License. 
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see           
 * <http://www.gnu.org/licenses/>.
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */

package org.jumpmind.symmetric.service.impl;

import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import org.apache.commons.lang.time.DateUtils;
import org.jumpmind.db.sql.ISqlRowMapper;
import org.jumpmind.db.sql.Row;
import org.jumpmind.symmetric.common.ParameterConstants;
import org.jumpmind.symmetric.db.ISymmetricDialect;
import org.jumpmind.symmetric.model.DataGap;
import org.jumpmind.symmetric.model.IncomingBatch;
import org.jumpmind.symmetric.model.OutgoingBatch;
import org.jumpmind.symmetric.service.ClusterConstants;
import org.jumpmind.symmetric.service.IClusterService;
import org.jumpmind.symmetric.service.IParameterService;
import org.jumpmind.symmetric.service.IPurgeService;
import org.jumpmind.symmetric.statistic.IStatisticManager;

/**
 * @see IPurgeService
 */
public class PurgeService extends AbstractService implements IPurgeService {

    enum MinMaxDeleteSql {
        DATA, DATA_EVENT, OUTGOING_BATCH, STRANDED_DATA
    };

    private IClusterService clusterService;

    private IStatisticManager statisticManager;

    public PurgeService(IParameterService parameterService, ISymmetricDialect symmetricDialect,
            IClusterService clusterService, IStatisticManager statisticManager) {
        super(parameterService, symmetricDialect);
        this.clusterService = clusterService;
        this.statisticManager = statisticManager;
        setSqlMap(new PurgeServiceSqlMap(symmetricDialect.getPlatform(),
                createSqlReplacementTokens()));
    }

    public long purgeOutgoing() {
        long rowsPurged = 0;
        Calendar retentionCutoff = Calendar.getInstance();
        retentionCutoff.add(Calendar.MINUTE,
                -parameterService.getInt(ParameterConstants.PURGE_RETENTION_MINUTES));
        rowsPurged += purgeOutgoing(retentionCutoff);
        return rowsPurged;
    }

    public long purgeIncoming() {
        long rowsPurged = 0;
        Calendar retentionCutoff = Calendar.getInstance();
        retentionCutoff.add(Calendar.MINUTE,
                -parameterService.getInt(ParameterConstants.PURGE_RETENTION_MINUTES));
        rowsPurged += purgeIncoming(retentionCutoff);
        return rowsPurged;
    }

    public long purgeDataGaps() {
        long rowsPurged = 0;
        Calendar retentionCutoff = Calendar.getInstance();
        retentionCutoff.add(Calendar.MINUTE, -parameterService
                .getInt(ParameterConstants.ROUTING_DATA_READER_TYPE_GAP_RETENTION_MINUTES));
        rowsPurged += purgeDataGaps(retentionCutoff);
        return rowsPurged;
    }

    public long purgeDataGaps(Calendar retentionCutoff) {
        long rowsPurged = -1l;
        try {
            if (clusterService.lock(ClusterConstants.PURGE_DATA_GAPS)) {
                try {
                    log.info("The data gap purge process is about to run");
                    rowsPurged = sqlTemplate.update(getSql("deleteFromDataGapsSql"),
                            new Object[] { retentionCutoff.getTime(), DataGap.Status.GP.name() });
                    log.info("Purged {} data gap rows", rowsPurged);
                } finally {
                    clusterService.unlock(ClusterConstants.PURGE_DATA_GAPS);
                    log.info("The data gap purge process has completed");
                }

            } else {
                log.warn("Did not run the data gap purge process because the cluster service has it locked");
            }
        } catch (Exception ex) {
            log.error(ex.getMessage(), ex);
        }
        return rowsPurged;
    }

    public long purgeOutgoing(Calendar retentionCutoff) {
        long rowsPurged = 0;
        try {
            if (clusterService.lock(ClusterConstants.PURGE_OUTGOING)) {
                try {
                    log.info("The outgoing purge process is about to run for data older than {}",
                            SimpleDateFormat.getDateTimeInstance()
                                    .format(retentionCutoff.getTime()));
                    rowsPurged += purgeStrandedBatches();
                    rowsPurged += purgeDataRows(retentionCutoff);
                    rowsPurged += purgeOutgoingBatch(retentionCutoff);
                } finally {
                    clusterService.unlock(ClusterConstants.PURGE_OUTGOING);
                    log.info("The outgoing purge process has completed");
                }
            } else {
                log.info("Could not get a lock to run an outgoing purge");
            }
        } catch (Exception ex) {
            log.error(ex.getMessage(), ex);
        }
        return rowsPurged;
    }

    private long purgeOutgoingBatch(final Calendar time) {
        log.info("Getting range for outgoing batch");
        long[] minMax = queryForMinMax(getSql("selectOutgoingBatchRangeSql"),
                new Object[] { time.getTime(), OutgoingBatch.Status.OK.name() });
        int maxNumOfBatchIdsToPurgeInTx = parameterService
                .getInt(ParameterConstants.PURGE_MAX_NUMBER_OF_BATCH_IDS);
        int maxNumOfDataEventsToPurgeInTx = parameterService
                .getInt(ParameterConstants.PURGE_MAX_NUMBER_OF_EVENT_BATCH_IDS);
        int dataEventsPurgedCount = purgeByMinMax(minMax, MinMaxDeleteSql.DATA_EVENT,
                time.getTime(), maxNumOfDataEventsToPurgeInTx);
        statisticManager.incrementPurgedDataEventRows(dataEventsPurgedCount);
        int outgoingbatchPurgedCount = purgeByMinMax(minMax, MinMaxDeleteSql.OUTGOING_BATCH,
                time.getTime(), maxNumOfBatchIdsToPurgeInTx);
        statisticManager.incrementPurgedBatchOutgoingRows(outgoingbatchPurgedCount);
        return dataEventsPurgedCount + outgoingbatchPurgedCount;
    }

    private long purgeStrandedBatches() {
        int updateStrandedBatchesCount = sqlTemplate.update(getSql("updateStrandedBatches"),
                OutgoingBatch.Status.OK.name(), 1, OutgoingBatch.Status.OK.name());
        if (updateStrandedBatchesCount > 0) {
            log.info(
                    "Set the status to {} for {} batches that no longer are associated with valid nodes",
                    OutgoingBatch.Status.OK.name(), updateStrandedBatchesCount);
            statisticManager.incrementPurgedBatchOutgoingRows(updateStrandedBatchesCount);
        }
        return updateStrandedBatchesCount;
    }

    private long purgeDataRows(final Calendar time) {
        log.info("Getting range for data");
        long[] minMax = queryForMinMax(getSql("selectDataRangeSql"), new Object[0]);
        int maxNumOfDataIdsToPurgeInTx = parameterService
                .getInt(ParameterConstants.PURGE_MAX_NUMBER_OF_DATA_IDS);
        int dataDeletedCount = purgeByMinMax(minMax, MinMaxDeleteSql.DATA, time.getTime(),
                maxNumOfDataIdsToPurgeInTx);
        statisticManager.incrementPurgedDataRows(dataDeletedCount);
        int strandedDeletedCount = purgeByMinMax(minMax, MinMaxDeleteSql.STRANDED_DATA,
                time.getTime(), maxNumOfDataIdsToPurgeInTx);
        statisticManager.incrementPurgedDataRows(strandedDeletedCount);
        return dataDeletedCount + strandedDeletedCount;

    }

    private long[] queryForMinMax(String sql, Object... params) {
        long[] minMax = sqlTemplate.queryForObject(sql, new ISqlRowMapper<long[]>() {
            public long[] mapRow(Row rs) {
                return new long[] { rs.getLong("min_id"), rs.getLong("max_id") };
            }
        }, params);
        return minMax;
    }

    private int purgeByMinMax(long[] minMax, MinMaxDeleteSql identifier, Date retentionTime,
            int maxNumtoPurgeinTx) {
        long minId = minMax[0];
        long purgeUpToId = minMax[1];
        long ts = System.currentTimeMillis();
        int totalCount = 0;
        int totalDeleteStmts = 0;
        Timestamp cutoffTime = new Timestamp(retentionTime.getTime());
        log.info("About to purge {}", identifier.toString().toLowerCase());
        while (minId <= purgeUpToId) {
            totalDeleteStmts++;
            long maxId = minId + maxNumtoPurgeinTx;
            if (maxId > purgeUpToId) {
                maxId = purgeUpToId;
            }

            String deleteSql = null;
            Object[] args = null;

            switch (identifier) {
                case DATA:
                    deleteSql = getSql("deleteDataSql");
                    args = new Object[] { minId, maxId, cutoffTime, minId, maxId, minId, maxId, OutgoingBatch.Status.OK.name() };
                    break;
                case DATA_EVENT:
                    deleteSql = getSql("deleteDataEventSql");
                    args = new Object[] { minId, maxId, OutgoingBatch.Status.OK.name(), minId,
                            maxId };
                    break;
                case OUTGOING_BATCH:
                    deleteSql = getSql("deleteOutgoingBatchSql");
                    args = new Object[] { OutgoingBatch.Status.OK.name(), minId, maxId, minId,
                            maxId };
                    break;
                case STRANDED_DATA:
                    deleteSql = getSql("deleteStrandedData");
                    args = new Object[] { minId, maxId, cutoffTime, minId, maxId };
                    break;
            }

            totalCount += sqlTemplate.update(deleteSql, args);

            if (totalCount > 0
                    && (System.currentTimeMillis() - ts > DateUtils.MILLIS_PER_MINUTE * 5)) {
                log.info("Purged {} of {} rows so far using {} statements", new Object[] {
                        totalCount, identifier.toString().toLowerCase(), totalDeleteStmts });
                ts = System.currentTimeMillis();
            }
            minId = maxId + 1;
        }
        log.info("Done purging {} of {} rows", totalCount, identifier.toString().toLowerCase());
        return totalCount;
    }

    public long purgeIncoming(Calendar retentionCutoff) {
        long purgedRowCount = 0;
        try {
            if (clusterService.lock(ClusterConstants.PURGE_INCOMING)) {
                try {
                    log.info("The incoming purge process is about to run");
                    purgedRowCount = purgeIncomingBatch(retentionCutoff);
                } finally {
                    clusterService.unlock(ClusterConstants.PURGE_INCOMING);
                    log.info("The incoming purge process has completed");
                }
            } else {
                log.info("Could not get a lock to run an incoming purge");
            }
        } catch (Exception ex) {
            log.error(ex.getMessage(), ex);
        }
        return purgedRowCount;
    }

    private long purgeIncomingBatch(final Calendar time) {
        log.info("Getting range for incoming batch");
        List<NodeBatchRange> nodeBatchRangeList = sqlTemplate.query(
                getSql("selectIncomingBatchRangeSql"), new ISqlRowMapper<NodeBatchRange>() {
                    public NodeBatchRange mapRow(Row rs) {
                        return new NodeBatchRange(rs.getString("node_id"), rs.getLong("min_id"), rs
                                .getLong("max_id"));
                    }
                }, time.getTime(), IncomingBatch.Status.OK.name());
        int incomingBatchesPurgedCount = purgeByNodeBatchRangeList(nodeBatchRangeList);
        statisticManager.incrementPurgedBatchIncomingRows(incomingBatchesPurgedCount);
        return incomingBatchesPurgedCount;
    }

    private int purgeByNodeBatchRangeList(List<NodeBatchRange> nodeBatchRangeList) {
        long ts = System.currentTimeMillis();
        int totalCount = 0;
        int totalDeleteStmts = 0;
        log.info("About to purge incoming batch");

        for (NodeBatchRange nodeBatchRange : nodeBatchRangeList) {
            int maxNumOfDataIdsToPurgeInTx = parameterService
                    .getInt(ParameterConstants.PURGE_MAX_NUMBER_OF_BATCH_IDS);
            long minBatchId = nodeBatchRange.getMinBatchId();
            long purgeUpToBatchId = nodeBatchRange.getMaxBatchId();

            while (minBatchId <= purgeUpToBatchId) {
                totalDeleteStmts++;
                long maxBatchId = minBatchId + maxNumOfDataIdsToPurgeInTx;
                if (maxBatchId > purgeUpToBatchId) {
                    maxBatchId = purgeUpToBatchId;
                }
                totalCount += sqlTemplate.update(getSql("deleteIncomingBatchSql"), new Object[] { minBatchId, maxBatchId,
                        nodeBatchRange.getNodeId(), IncomingBatch.Status.OK.name() });
                minBatchId = maxBatchId + 1;
            }

            if (totalCount > 0
                    && (System.currentTimeMillis() - ts > DateUtils.MILLIS_PER_MINUTE * 5)) {
                log.info("Purged {} incoming batch rows so far using {} statements", new Object[] {
                        totalCount, totalDeleteStmts });
                ts = System.currentTimeMillis();
            }
        }
        log.info("Done purging {} incoming batch rows", totalCount);
        return totalCount;
    }

    class NodeBatchRange {
        private String nodeId;

        private long minBatchId;

        private long maxBatchId;

        public NodeBatchRange(String nodeId, long minBatchId, long maxBatchId) {
            this.nodeId = nodeId;
            this.minBatchId = minBatchId;
            this.maxBatchId = maxBatchId;
        }

        public String getNodeId() {
            return nodeId;
        }

        public long getMaxBatchId() {
            return maxBatchId;
        }

        public long getMinBatchId() {
            return minBatchId;
        }
    }

    public void purgeAllIncomingEventsForNode(String nodeId) {
        int count = sqlTemplate.update(getSql("deleteIncomingBatchByNodeSql"),
                new Object[] { nodeId });
        log.info("Purged all {} incoming batch for node {}", count, nodeId);
    }

}

File: symmetric-client-clib/src/service/PurgeService.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "service/PurgeService.h"

typedef struct SymMinMax {
    long min;
    long max;
} SymMinMax;

typedef struct SymNodeBatchRange {
    char *nodeId;
    long minBatchId;
    long maxBatchId;
} SymNodeBatchRange;

static SymMinMax * SymPurgeService_minMaxMapper(SymRow *row) {
    long min = row->getLong(row, "min_id");
    long max = row->getLong(row, "max_id");

    SymMinMax *minMax = (SymMinMax *) calloc(1, sizeof(SymMinMax));
    minMax->min = min;
    minMax->max = max;
    return minMax;
}

static SymNodeBatchRange * SymPurgeService_nodeBatchRangeMapper(SymRow *row) {
    SymNodeBatchRange *nodeBatchRange = (SymNodeBatchRange *) calloc(1, sizeof(SymNodeBatchRange));
    nodeBatchRange->nodeId = row->getStringNew(row, "node_id");
    nodeBatchRange->minBatchId = row->getLong(row, "min_id");
    nodeBatchRange->maxBatchId = row->getLong(row, "max_id");
    return nodeBatchRange;
}

SymMinMax * SymPurgeService_queryForMinMax(SymPurgeService *this, char *sql, SymStringArray *sqlParams) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    int error;
    SymList *minMaxList = sqlTemplate->query(sqlTemplate, sql, sqlParams, NULL, &error, (void *) SymPurgeService_minMaxMapper);
    if (minMaxList->size > 0) {
        return minMaxList->get(minMaxList, 0);
    } else {
        return NULL;
    }
}

int SymPurgeService_purgeByMinMax(SymPurgeService *this, SymMinMax *minMax, SymMinMaxDeleteSql identifier, SymDate *retentionTime, int maxNumtoPurgeinTx) {

    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);

    long minId = minMax->min;
    long purgeUpToId = minMax->max;

    int totalCount = 0;
    int totalDeleteStmts = 0;

    SymDate *cutoffTime = retentionTime;

    SymLog_info("About to purge %d", identifier);

    while (minId <= purgeUpToId) {
        totalDeleteStmts++;
        long maxId = minId + maxNumtoPurgeinTx;
        if (maxId > purgeUpToId) {
            maxId = purgeUpToId;
        }

        char* deleteSql = NULL;
        char *minIdString = SymStringUtils_format("%ld", minId);
        char *maxIdString = SymStringUtils_format("%ld", maxId);
        SymStringArray *args = SymStringArray_new(NULL);

        switch (identifier) {
        case SYM_MIN_MAX_DELETE_DATA:
            deleteSql = SYM_SQL_DELETE_DATA;
            args->add(args, minIdString);
            args->add(args, maxIdString);
            args->add(args, cutoffTime->dateTimeString);
            args->add(args, minIdString);
            args->add(args, maxIdString);
            args->add(args, minIdString);
            args->add(args, maxIdString);
            args->add(args, SYM_OUTGOING_BATCH_OK);
            break;
        case SYM_MIN_MAX_DELETE_DATA_EVENT:
            deleteSql = SYM_SQL_DELETE_DATA_EVENT;
            args->add(args, minIdString);
            args->add(args, maxIdString);
            args->add(args, SYM_OUTGOING_BATCH_OK);
            args->add(args, minIdString);
            args->add(args, maxIdString);
            break;
        case SYM_MIN_MAX_DELETE_OUTGOING_BATCH:
            deleteSql = SYM_SQL_DELETE_OUTGOING_BATCH;
            args->add(args, SYM_OUTGOING_BATCH_OK);
            args->add(args, minIdString);
            args->add(args, maxIdString);
            args->add(args, minIdString);
            args->add(args, maxIdString);
            break;
        case SYM_MIN_MAX_DELETE_STRANDED_DATA:
            deleteSql = SYM_SQL_DELETE_STRANDED_DATA;
            args->add(args, minIdString);
            args->add(args, maxIdString);
            args->add(args, cutoffTime->dateTimeString);
            args->add(args, minIdString);
            args->add(args, maxIdString);
            break;
        }

        char *argsString = args->toString(args);
        SymLog_debug("Running the following statement: %s with the following arguments: %s", deleteSql, argsString);
        free(argsString);

        int error;
        int count = sqlTemplate->update(sqlTemplate, deleteSql, args, NULL, &error);
        totalCount += count;

        SymLog_info("Purged %d of (identifier = %d) rows so far using %d statements",
                totalCount, identifier, totalDeleteStmts );

        minId = maxId + 1;
        args->destroy(args);
        free(minIdString);
        free(maxIdString);
    }

    return totalCount;
}

long SymPurgeService_purgeStrandedBatches(SymPurgeService *this) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);

    SymStringArray *args = SymStringArray_new(NULL);
    args->add(args, SYM_OUTGOING_BATCH_OK);
    args->add(args, "1");
    args->add(args, SYM_OUTGOING_BATCH_OK);

    int error;
    int rowsAffected = sqlTemplate->update(sqlTemplate, SYM_SQL_UPDATE_STRANDED_BATCHES, args, NULL, &error);

    args->destroy(args);
    return rowsAffected;
}

long SymPurgeService_purgeDataRows(SymPurgeService *this, SymDate *time) {
    SymLog_info("Getting range for data");
    SymMinMax *minMax = SymPurgeService_queryForMinMax(this, SYM_SQL_SELECT_DATA_RANGE, NULL);
    int maxNumOfDataIdsToPurgeInTx =
            this->parameterService->getInt(this->parameterService, SYM_PARAMETER_PURGE_MAX_NUMBER_OF_DATA_IDS, 5000);
    int dataDeletedCount = SymPurgeService_purgeByMinMax(this, minMax,
            SYM_MIN_MAX_DELETE_DATA, time, maxNumOfDataIdsToPurgeInTx);
    int strandedDeletedCount = SymPurgeService_purgeByMinMax(this, minMax,
            SYM_MIN_MAX_DELETE_STRANDED_DATA, time, maxNumOfDataIdsToPurgeInTx);

    return dataDeletedCount + strandedDeletedCount;
}

long SymPurgeService_purgeOutgoingBatch(SymPurgeService *this, SymDate *time) {
    SymLog_info("Getting range for outgoing batch");

    SymStringArray *args = SymStringArray_new(NULL);
    args->add(args, time->dateTimeString);
    args->add(args, SYM_OUTGOING_BATCH_OK);
    SymMinMax *minMax = SymPurgeService_queryForMinMax(this, SYM_SQL_SELECT_OUTGOING_BATCH_RANGE, args);
    args->destroy(args);

    int maxNumOfBatchIdsToPurgeInTx =
            this->parameterService->getInt(this->parameterService, SYM_PARAMETER_PURGE_MAX_NUMBER_OF_BATCH_IDS, 5000);
    int maxNumOfDataEventsToPurgeInTx =
            this->parameterService->getInt(this->parameterService, SYM_PARAMETER_PURGE_MAX_NUMBER_OF_EVENT_BATCH_IDS, 5);

    int dataEventsPurgedCount = SymPurgeService_purgeByMinMax(this, minMax, SYM_MIN_MAX_DELETE_DATA_EVENT,
            time, maxNumOfDataEventsToPurgeInTx);
    int outgoingbatchPurgedCount = SymPurgeService_purgeByMinMax(this, minMax, SYM_MIN_MAX_DELETE_OUTGOING_BATCH,
            time, maxNumOfBatchIdsToPurgeInTx);
    return dataEventsPurgedCount + outgoingbatchPurgedCount;
}

int SymPurgeService_purgeByNodeBatchRangeList(SymPurgeService *this, SymList *nodeBatchRangeList) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);

    int totalCount = 0;
    int totalDeleteStmts = 0;
    SymLog_info("About to purge incoming batch");

    int i;
    for (i = 0; i < nodeBatchRangeList->size; ++i) {
        SymNodeBatchRange *nodeBatchRange = nodeBatchRangeList->get(nodeBatchRangeList, i);
        int maxNumOfBatchIdsToPurgeInTx =
                this->parameterService->getInt(this->parameterService, SYM_PARAMETER_PURGE_MAX_NUMBER_OF_BATCH_IDS, 5000);

        char *minBatchIdString = SymStringUtils_format("%ld", nodeBatchRange->minBatchId);
        char *maxBatchIdString = SymStringUtils_format("%ld", nodeBatchRange->maxBatchId);

        long minBatchId = nodeBatchRange->minBatchId;
        long purgeUpToBatchId = nodeBatchRange->maxBatchId;

        while (minBatchId <= purgeUpToBatchId) {
            totalDeleteStmts++;
            long maxBatchId = minBatchId + maxNumOfBatchIdsToPurgeInTx;
            if (maxBatchId > purgeUpToBatchId) {
                maxBatchId = purgeUpToBatchId;
            }
            SymStringArray *args = SymStringArray_new(NULL);
            args->add(args, minBatchIdString);
            args->add(args, maxBatchIdString);
            args->add(args, nodeBatchRange->nodeId);
            args->add(args, SYM_OUTGOING_BATCH_OK);

            int error;
            totalCount +=  sqlTemplate->update(sqlTemplate, SYM_SQL_PURGE_INCOMING_BATCH, args, NULL, &error);
            args->destroy(args);
            minBatchId = maxBatchId + 1;
        }

        SymLog_info("Purged %d incoming batch rows so far using %d statements",
                totalCount, totalDeleteStmts );
    }
    SymLog_info("Done purging %d incoming batch rows", totalCount);
    return totalCount;
}

long SymPurgeService_purgeIncomingBatch(SymPurgeService *this, SymDate *time) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);

    SymStringArray *args = SymStringArray_new(NULL);
    args->add(args, time->dateTimeString);
    args->add(args, SYM_OUTGOING_BATCH_OK);

    SymLog_info("Getting range for incoming batch");
    int error;
    SymList *nodeBatchRangeList = sqlTemplate->query(sqlTemplate, SYM_SQL_SELECT_INCOMING_BATCH_RANGE,
            args, NULL, &error, (void *) SymPurgeService_nodeBatchRangeMapper);
    args->destroy(args);

    int incomingBatchesPurgedCount = SymPurgeService_purgeByNodeBatchRangeList(this, nodeBatchRangeList);
    nodeBatchRangeList->destroy(nodeBatchRangeList);

    return incomingBatchesPurgedCount;
}

long SymPurgeService_purgeIncomingError(SymPurgeService *this) {
    // TODO: support purge incoming error?
//    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
//    SymLog_info("Purging incoming error rows");
//    int error;
//    long rowCount = sqlTemplate.update(getSql("deleteIncomingErrorsSql"));
//    sqlTemplate->update(sqlTemplate, SYM_SQL_DELETE_INCOMING_BATCH, NULL, NULL, &error);
//    SymLog_info("Purged %ld incoming error rows", rowCount);
//    return rowCount;

    return 0;
}

long SymPurgeService_purgeIncomingBeforeDate(SymPurgeService *this, SymDate *retentionCutoff) {
    long purgedRowCount = 0;
    SymLog_info("The incoming purge process is about to run");
    purgedRowCount = SymPurgeService_purgeIncomingBatch(this, retentionCutoff);
    purgedRowCount += SymPurgeService_purgeIncomingError(this);
    SymLog_info("The incoming purge process has completed");

    return purgedRowCount;
}

long SymPurgeService_purgeOutgoingBeforeDate(SymPurgeService *this, SymDate *retentionCutoff) {
    long rowsPurged = 0;
    SymLog_info("The outgoing purge process is about to run for data older than %s",
            retentionCutoff->dateTimeString);
    rowsPurged += SymPurgeService_purgeStrandedBatches(this);
    rowsPurged += SymPurgeService_purgeDataRows(this, retentionCutoff);
    rowsPurged += SymPurgeService_purgeOutgoingBatch(this, retentionCutoff);
    SymLog_info("The outgoing purge process has completed.");

    return rowsPurged;
}

SymDate *SymPurgeService_getRetentionCutoff(SymPurgeService *this) {

    time_t timeInSeconds = time(NULL);
    long timeInMinutes = timeInSeconds / 60;

    int purgeRetentionMinutes =
            this->parameterService->getInt(this->parameterService, SYM_PARAMETER_PURGE_RETENTION_MINUTES, 1440);
    long retentionCutoffMinutes = timeInMinutes - purgeRetentionMinutes;

    SymDate *retentionCutoff = SymDate_newWithTime(retentionCutoffMinutes*60);
    return retentionCutoff;
}

long SymPurgeService_purgeIncoming(SymPurgeService *this) {
    long rowsPurged = 0;
    SymDate *retentionCutoff = SymPurgeService_getRetentionCutoff(this);
    rowsPurged += SymPurgeService_purgeIncomingBeforeDate(this, retentionCutoff);
    retentionCutoff->destroy(retentionCutoff);
    return rowsPurged;
}

long SymPurgeService_purgeOutgoing(SymPurgeService *this) {
    long rowsPurged = 0;
    SymDate *retentionCutoff = SymPurgeService_getRetentionCutoff(this);
    rowsPurged += SymPurgeService_purgeOutgoingBeforeDate(this, retentionCutoff);
    retentionCutoff->destroy(retentionCutoff);
    return rowsPurged;
}

void SymPurgeService_purgeAllIncomingEventsForNode(SymPurgeService *this, char *nodeId) {
    // TODO
}

void SymPurgeService_destroy(SymPurgeService *this) {
    free(this);
}

SymPurgeService * SymPurgeService_new(SymPurgeService *this, SymParameterService *parameterService,
        SymDialect *symmetricDialect, SymDatabasePlatform *platform) {
    if (this == NULL) {
        this = (SymPurgeService *) calloc(1, sizeof(SymPurgeService));
    }
    this->parameterService = parameterService;
    this->symmetricDialect = symmetricDialect;
    this->platform = platform;

    this->purgeIncoming = (void *) &SymPurgeService_purgeIncoming;
    this->purgeOutgoing = (void *) &SymPurgeService_purgeOutgoing;
    this->purgeIncomingBeforeDate = (void *) &SymPurgeService_purgeIncomingBeforeDate;
    this->purgeOutgoingBeforeDate = (void *) &SymPurgeService_purgeOutgoingBeforeDate;

    this->destroy = (void *) &SymPurgeService_destroy;
    return this;
}
